statsmodels/tsa/stattools.py

```diff
@@ -1735,6 +1735,282 @@ def _sigma_est_kpss(resids, nobs, lags):
+        investigation. Journal of Business & Economic Statistics, 7: 147-159.
+
+        Zivot, E., and Andrews, D.W.K. (1992). Further evidence on the great
+        crash, the oil-price shock, and the unit-root hypothesis. Journal of
+        Business & Economic Studies, 10: 251-270.
+        """
+        if regression not in ['c', 't', 'ct']:
+            raise ValueError(
+                'ZA: regression option \'%s\' not understood' % regression)
+        if not isinstance(trim, float) or trim < 0 or trim > (1. / 3.):
+            raise ValueError(
+                'ZA: trim value must be a float in range [0, 0.333]')
+        x = np.asarray(x)
+        if x.ndim > 2 or (x.ndim == 2 and x.shape[1] != 1):
+            raise ValueError(
+                'ZA: x must be a 1d array or a 2d array with a single column')
+        x = np.reshape(x, (-1, 1))
+        nobs = x.shape[0]
+        if autolag:
+            baselags = adfuller(x[:, 0], maxlag=maxlag, regression='ct',
+                                autolag=autolag)[2]
+        elif maxlag:
+            baselags = maxlag
+        else:
+            baselags = int(12. * np.power(nobs / 100., 1 / 4.))
+        trimcnt = int(nobs * trim)
+        start_period = trimcnt
+        end_period = nobs - trimcnt
+        if regression == 'ct':
+            basecols = 5
+        else:
+            basecols = 4
+        # first-diff y and standardize for numerical stability
+        dy = np.diff(x, axis=0)[:, 0]
+        dy /= np.sqrt(dy.T.dot(dy))
+        x = x / np.sqrt(x.T.dot(x))
+        # reserve exog space
+        exog = np.zeros((dy[baselags:].shape[0], basecols + baselags))
+        # normalize constant for stability in long time series
+        c_const = 1 / np.sqrt(nobs)  # Normalize
+        exog[:, 0] = c_const
+        # lagged y and dy
+        exog[:, basecols - 1] = x[baselags:(nobs - 1), 0]
+        exog[:, basecols:] = lagmat(
+            dy, baselags, trim='none')[baselags:exog.shape[0] + baselags]
+        # better time trend: t_const @ t_const = 1 for large nobs
+        t_const = np.arange(1.0, nobs + 2)
+        t_const *= np.sqrt(3) / nobs ** (3 / 2)
+        # iterate through the time periods
+        stats = np.full(end_period + 1, np.inf)
+        for bp in range(start_period + 1, end_period + 1):
+            # update intercept dummy / trend / trend dummy
+            cutoff = (bp - (baselags + 1))
+            if regression != 't':
+                exog[:cutoff, 1] = 0
+                exog[cutoff:, 1] = c_const
+                exog[:, 2] = t_const[(baselags + 2):(nobs + 1)]
+                if regression == 'ct':
+                    exog[:cutoff, 3] = 0
+                    exog[cutoff:, 3] = t_const[1:(nobs - bp + 1)]
+            else:
+                exog[:, 1] = t_const[(baselags + 2):(nobs + 1)]
+                exog[:(cutoff-1), 2] = 0
+                exog[(cutoff-1):, 2] = t_const[0:(nobs - bp + 1)]
+            # check exog rank on first iteration
```

Comment:
this method is doing a lot.  are there any reasonable sub-scopes it could be split into?

Reply:
Split off the endog/exog creation/updating into two separate helper functions.